File: PC Por camara/PC por Camara.cpp
#include <opencv2/opencv.hpp>
#include <iostream>
#include <cmath>
#include <algorithm>

using namespace std;
using namespace cv;

int main(int argc, char* argv[])
{
    // Inicializar la cámara
    VideoCapture camera(0);
    if (!camera.isOpened()) {
        cerr << "ERROR: No se pudo abrir la cámara." << endl;
        return 1;
    }

    camera.set(CAP_PROP_FRAME_WIDTH, 640);
    camera.set(CAP_PROP_FRAME_HEIGHT, 480);

    // Cargar el clasificador de cascada de Haar para la mano
    CascadeClassifier hand_cascade;
    if (!hand_cascade.load("haarcascade_hand.xml")) {
        cerr << "ERROR: No se pudo cargar el clasificador de cascada de Haar." << endl;
        return 1;
    }

    // Definir las variables para los límites del área del pecho
    int chest_x_min = 100;
    int chest_x_max = 200;
    int chest_y_min = 100;
    int chest_y_max = 200;

    bool pyautogui_failsafe = true;

    while (true)
    {
        // Capturar una imagen de la cámara
        Mat frame;
        if (!camera.read(frame)) {
            cerr << "ERROR: No se pudo leer un frame de la cámara." << endl;
            break;
        }
        resize(frame, frame, Size(0, 0), 0.5, 0.5);

        // Convertir la imagen a escala de grises
        Mat gray;
        cvtColor(frame, gray, COLOR_BGR2GRAY);

        // Detectar movimientos de mano en la imagen
        vector<Rect> hands;
        hand_cascade.detectMultiScale(gray, hands);

        // Recorrer la lista de manos detectadas
        for (size_t i = 0; i < hands.size(); i++)
        {
            int x = hands[i].x;
            int y = hands[i].y;
            int w = hands[i].width;
            int h = hands[i].height;

            // Si la mano derecha está sobre el área del pecho
            // y el modo FAILSAFE está deshabilitado, habilitar el modo FAILSAFE
            if (x > chest_x_min && x < chest_x_max && y > chest_y_min && y < chest_y_max && !pyautogui_failsafe) {
                pyautogui_failsafe = true;
            }

            // Si la mano izquierda está sobre el área del pecho
            // y el modo FAILSAFE está habilitado, deshabilitar el modo FAILSAFE
            else if (x > chest_x_min && x < chest_x_max && y > chest_y_min && y < chest_y_max && pyautogui_failsafe) {
                pyautogui_failsafe = false;
            }

            // Mover el cursor del ratón a la posición de la mano
            // si el modo FAILSAFE está deshabilitado
            if (!pyautogui_failsafe) {
                SetCursorPos(x, y);
            }
        }

        // Si se han detectado dos manos
        if (hands.size() == 2)
        {
            // Obtener la posición de las dos manos
            int x1 = hands[0].x;
            int y1 = hands[0].y;
            int w1 = hands[0].width;
            int h1 = hands[0].height;
            int x2 = hands[1].x;
            int y2 = hands[1].y;
            int w2 = hands[1].width;
            int h2 = hands[1].height;

            // Si ambas manos están levantadas a la altura del pecho
            // y el modo FAILSAFE está deshabilitado, escribir en el teclado virtual
            if (y1 > chest_y_min && y1 < chest_y_max && y2 > chest_y_min && y2 < chest_y_max && !pyautogui_failsafe)
            {
                // Escribir en el teclado virtual
                keybd_event(VkKeyScan('H'), 0, 0, 0);
                keybd_event(VkKeyScan('H'), 0, KEYEVENTF_KEYUP, 0);
                keybd_event(VkKeyScan('E'), 0, 0, 0);
                keybd_event(VkKeyScan('E'), 0, KEYEVENTF_KEYUP, 0);
                keybd_event(VkKeyScan('L'), 0, 0, 0);
                keybd_event(VkKeyScan('L'), 0, KEYEVENTF_KEYUP, 0);
                keybd_event(VkKeyScan('L'), 0, 0, 0);
                keybd_event(VkKeyScan('L'), 0, KEYEVENTF_KEYUP, 0);
                keybd_event(VkKeyScan('O'), 0, 0, 0);
                keybd_event(VkKeyScan('O'), 0, KEYEVENTF_KEYUP, 0);
            }
        }

        // Mostrar la imagen de la cámara
        imshow("Camera", frame);

        // Si se ha pulsado la tecla 'q', salir del bucle
        if (waitKey(1) == 'q') {
            break;
        }
    }

    return 0;
}

File: PC Por camara/PC por Camara.py
# Importar las librerías necesarias
import cv2
import pyautogui

# Inicializar la cámara
camera = cv2.VideoCapture(0)
camera.set(cv2.CAP_PROP_FRAME_WIDTH, 640)
camera.set(cv2.CAP_PROP_FRAME_HEIGHT, 480)


# Verificar que la cámara se haya abierto correctamente
if not camera.isOpened():
    print("No se pudo abrir la cámara.")
    exit()

# Cargar el clasificador de cascada de Haar para la mano
hand_cascade = cv2.CascadeClassifier('haarcascade_hand.xml')

# Verificar que se haya cargado el clasificador correctamente
if hand_cascade.empty():
    print("No se pudo cargar el clasificador de cascada de Haar.")
    exit()

# Definir las variables para los límites del área del pecho
chest_x_min = 100
chest_x_max = 200
chest_y_min = 100
chest_y_max = 200

# Habilitar el modo FAILSAFE de PyAutoGUI
pyautogui.FAILSAFE = True

while True:
    # Capturar una imagen de la cámara
    _, frame = camera.read()
    frame = cv2.resize(frame, (0, 0), fx=0.5, fy=0.5)


    # Convertir la imagen a escala de grises
    gray = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)

    # Detectar movimientos de mano en la imagen
    hands = hand_cascade.detectMultiScale(gray)

    # Recorrer la lista de manos detectadas
    for (x, y, w, h) in hands:
        # Si la mano derecha está sobre el área del pecho
        # y el modo FAILSAFE está deshabilitado, habilitar el modo FAILSAFE
        if x > chest_x_min and x < chest_x_max and y > chest_y_min and y < chest_y_max and not pyautogui.FAILSAFE:
            pyautogui.FAILSAFE = True

        # Si la mano izquierda está sobre el área del pecho
        # y el modo FAILSAFE está habilitado, deshabilitar el modo FAILSAFE
        elif x > chest_x_min and x < chest_x_max and y > chest_y_min and y < chest_y_max and pyautogui.FAILSAFE:
            pyautogui.FAILSAFE = False

        # Mover el cursor del ratón a la posición de la mano
        # si el modo FAILSAFE está deshabilitado
        if not pyautogui.FAILSAFE:
            pyautogui.moveTo(x, y)

        # Si se han detectado dos manos
        if len(hands) == 2:
            # Obtener la posición de las dos manos
            x1, y1, w1, h1 = hands[0]
            x2, y2, w2, h2 = hands[1]

            # Si ambas manos están levantadas a la altura del pecho
            # y el modo FAILSAFE está deshabilitado, escribir en el teclado virtual
            if y1 > chest_y_min and y1 < chest_y_max and y2 > chest_y_min and y2 < chest_y_max and not pyautogui.FAILSAFE:
                # Escribir en el teclado virtual
                pyautogui.typewrite('Hola, estoy escribiendo con mis manos!')
            # Si ambas manos están levantadas a la altura del pecho
            # y están formando una cruz con los dedos índice, cerrar la funcionalidad
            elif y1 > chest_y_min and y1 < chest_y_max and y2 > chest_y_min and y2 < chest_y_max and x1 < x2 and y1 < y2:
                break
        # Mostrar la imagen en la ventana
        cv2.imshow('Webcam', frame)

        # Pausar el loop por un tiempo específico
        cv2.waitKey(30)

    # Salir del bucle si se pulsa la tecla 'q'
    if cv2.waitKey(1) & 0xFF == ord('q'):
        break

# Liberar la cámara y cerrar la ventana
camera.release()
cv2.destroyAllWindows()
